frontend/src/test/kea-test-utils/functions/toDispatchActionsInAnyOrder.ts

```diff
@@ -0,0 +1,85 @@
+import { BuiltLogic, LogicWrapper } from 'kea'
+import { ActionToDispatch, ExpectFunction, testUtilsContext } from '~/test/kea-test-utils'
+import { delay, objectsEqual } from 'lib/utils'
+import { waitForAction, waitForCondition } from 'kea-waitfor'
+
+const ASYNC_ACTION_WAIT_TIMEOUT = 3000
+
+export const toDispatchActionsInAnyOrder: ExpectFunction<ActionToDispatch[]> = {
+    common() {
+        testUtilsContext().ranActions = true
+    },
+
+    sync(logic, actions) {
+        const { notFound } = tryToSearchActions(logic, actions)
+        if (notFound.length > 0) {
+            return [{ operation: 'toDispatchActionsInAnyOrder', logic, payload: notFound }]
+        }
+    },
+
+    async async(logic, actions) {
+        const { notFound } = tryToSearchActions(logic, actions)
+        if (notFound.length > 0) {
+            await Promise.race([
+                delay(ASYNC_ACTION_WAIT_TIMEOUT).then(() => {
+                    throw new Error(
+                        `Timed out waiting for action: ${
+                            typeof notFound === 'object' ? JSON.stringify(notFound) : notFound
+                        } in logic ${logic?.pathString}`
+                    )
+                }),
+                Promise.all(
+                    notFound.map((act) =>
+                        typeof act === 'string'
+                            ? waitForAction(logic.actionTypes[act] || act)
+                            : typeof act === 'function'
+                            ? waitForCondition(act)
+                            : waitForCondition((a) => objectsEqual(a, act))
+                    )
+                ),
+            ])
+            // will not get called if the timeout throws above, otherwise it was found, and update the historyIndex
+            tryToSearchActions(logic, notFound)
+        }
+    },
+}
+
+interface NotFoundActions {
+    notFound: ActionToDispatch[]
+    lastIndex: number
+}
+
+function tryToSearchActions(logic: LogicWrapper | BuiltLogic, actions: ActionToDispatch[]): NotFoundActions {
+    const actionsToSearch = [...actions]
+    const { recordedHistory, historyIndex } = testUtilsContext()
+    const actionPointer = historyIndex || -1
+
+    const foundMap = new Map<ActionToDispatch, number>()
+    const alreadyFoundAtIndex = new Set<number>()
+
+    for (const action of actionsToSearch) {
+        for (let i = actionPointer + 1; i < recordedHistory.length; i++) {
+            const recordedAction = recordedHistory[i]
+            if (
+                (!alreadyFoundAtIndex.has(i) &&
```

Comment:
another nit...

there's a lot going on in these comparisons

I wonder if giving them each a name now helps someone in future if and when they come to change the conditional logic

e.g. not immediately apparent to me why we check the first one isn't present in alreadyFoundAtIndex but not the other two (whether that's naming or a comment or me being slow to understand :))

Reply:
Agreed that this is rather crap. The plan is to extract this `kea-test-utils` folder out into its own package next week and start adding proper tests to all these matchers... and this will probably get refactored once tests come on.

There's also a bigger issue with this matcher. The `sync` part above together with a notFound `async` might miss matching some events, as the pointer will be updated to after the last found sync event. It also doesn't pass `alreadyFoundAtIndex` from sync to async.

The idea behind `sync` and `async` above is that if we run:

```ts
expectLogic().toDispatchActions(['something'])
```

we don't need to add `await` if the action has already been called. This is just to make [some types of tests](https://github.com/PostHog/posthog/blob/ef70592603d316f20281cbe10d313d145c92736c/frontend/src/lib/components/TaxonomicFilter/infiniteListLogic.test.ts#L82-L114) even cleaner. 

We could simplify all of this by going `async` only, but the overhead of supporting both approaches is probably worth it here.